### PR TITLE
Handle non staking tokens

### LINKED ISSUE #1118: Withdraw unstakable tokens
It is possible that third parties may send tokens to the delegate account (for example, the WOSMO airdrop) that are unstakable. 

When tokens that do not match the zone.BaseDenom, or if LSM is enabled, does not match the valoper prefix (I believe a method for determining whether a token is a valid staking token exists already), then we should ICA MsgSend to the withdrawal account. 

From the withdrawal account, fees will be collected and the remainder will be distributed to the appropriate qAsset holders by existing logic.

### PR DESCRIPTION
**Pull request by joe-bowman**
## 1. Summary
Fixes #1118 - Allows deposits of non-staking tokens in deposit/delegate accounts to be collected and handled as rewards

## 2.Type of change

- [x] New feature (non-breaking change which adds functionality)

## 3. Implementation details

- Update ValidateCoinForZone to determine whether the deposit can be staked.
- Ensure Increment and Decrement of WithdrawalWaitgroup is done by helper methods to ensure no overflow.
- If coins cannot be staked, they are moved to the withdrawal account, where normal withdrawal processing can take place
- Fix race condition in non-staking denom rewards
- Fix identified race in un/re/delegation response

## Summary by CodeRabbit

- **New Features**
	- Introduced `IncrementWithdrawalWaitgroup` and `SetWithdrawalWaitgroup` methods for better withdrawal management.
	- Added `SendToWithdrawal` method for transferring coins in `keeper.go`.
	- Implemented `IsDepositAddress` method to validate deposit addresses.
- **Enhancements**
	- Improved logging and error handling across various modules.
	- Enhanced token validation logic in `ValidateCoinsForZone`.
	- Adjusted logging levels in `intent.go` and `withdrawal_record.go` for better clarity.
- **Bug Fixes**
	- Fixed error handling in `DepositLsmTxCallbackFailOnNonMatchingValidator` to prevent queue blocking.
	- Updated `HandleWithdrawRewards` function to use `SetWithdrawalWaitgroup` for consistency.
- **Refactor**
	- Refactored handling of `WithdrawalWaitgroup` using getter and setter methods across several files.
- **Documentation, Style, Tests, Chores**
	- Updated Go version requirement to 1.22 in Makefile and various workflow files.
	- Adjusted Dockerfiles to use updated Go and Alpine versions for improved security and compatibility.
- **Revert**
	- Not applicable in this release.

### CODE REVIEW

File: x/interchainstaking/keeper/keeper.go [hunk 1]
```diff
@@ -795,3 +795,18 @@ func (k *Keeper) UnmarshalValidator(data []byte) (lsmstakingtypes.Validator, err
...
+	sendMsg := banktypes.MsgSend{
+		FromAddress: sender.Address,
+		ToAddress:   zone.WithdrawalAddress.Address,
+		Amount:      amount,
+	}
+
+	msgs = append(msgs, &sendMsg)
+
+	return k.SubmitTx(ctx, msgs, sender, "", zone.MessagesPerTx)
+}
```
[the remark this answers is not in the document]
Comment (joe-bowman) [reply x/interchainstaking/keeper/keeper.go:812]: @coderabbitai - this code is functionally identical. 

in your code, is the reponse is nil, we return nil - if it is an error we return an error. 
in the original, we return the response.

File: x/interchainstaking/keeper/callbacks.go [hunk 2]
```diff
@@ -124,12 +124,12 @@ func RewardsCallback(k *Keeper, ctx sdk.Context, args []byte, query icqtypes.Que
 
 	// decrement waitgroup as we have received back the query
 	// (initially incremented in AfterEpochEnd)
-	err = zone.DecrementWithdrawalWaitgroup()
+	err = zone.DecrementWithdrawalWaitgroup(k.Logger(ctx), 1, "rewards callback")
```
[the remark this answers is not in the document]
Comment (joe-bowman) [reply x/interchainstaking/keeper/callbacks.go:127]: @coderabbitai DecrementWithdrawalWaitgroup handles underflow protection internal, that is the purpose of the helper

File: x/interchainstaking/keeper/callbacks.go [hunk 3]
```diff
@@ -622,7 +622,7 @@
 	}
 
 	k.Logger(ctx).Info("Received balance response for denom", "denom", coin.Denom)
-	err = zone.DecrementWithdrawalWaitgroup()
+	err = zone.DecrementWithdrawalWaitgroup(k.Logger(ctx), 1, "delegationaccountbalance callback")
```
[the remark this answers is not in the document]
Comment (joe-bowman) [reply x/interchainstaking/keeper/callbacks.go:625]: https://github.com/coderabbitai DecrementWithdrawalWaitgroup handles underflow protection internal, that is the purpose of the helper

File: x/interchainstaking/keeper/callbacks.go [hunk 4]
```diff
@@ -642,6 +642,12 @@
 
 	k.SetZone(ctx, &zone)
 
+	// if token is not valid for staking, then send to withdrawal account.
+	if valid, _ := zone.ValidateCoinsForZone(sdk.NewCoins(coin), k.GetValidatorAddressesAsMap(ctx, zone.ChainId)); !valid {
+		k.Logger(ctx).Info("token is not a valid staking token, so sending to withdrawal account for disbursal", "chain", zone.ChainId, "assets", coin)
+		return k.SendToWithdrawal(ctx, &zone, zone.DelegationAddress, sdk.NewCoins(coin))
+	}
+
```
[the remark this answers is not in the document]
Comment (joe-bowman) [reply x/interchainstaking/keeper/callbacks.go:650]: @coderabbitai if the denom is invalid, then the second argument returned by ValidateCoinsForZone (does the token match the list of validators provided) is irrelevant.

File: x/interchainstaking/keeper/receipt.go [hunk 5]
```diff
@@ -63,23 +63,32 @@
...
+
 	}
 	senderAccAddress, err := addressutils.AccAddressFromBech32(senderAddress, zone.GetAccountPrefix())
-	if err != nil {
+	if err != nil { // not sure this is ever reachable. A valid MsgSend must always have a valid sender.
 		k.Logger(ctx).Error("unable to decode sender address. Ignoring.", "senderAddress", senderAddress, "error", err)
 		k.NilReceipt(ctx, &zone, hash) // nil receipt will stop this hash being submitted again
-		return nil
+		return k.SendToWithdrawal(ctx, &zone, zone.DepositAddress, assets)
+
```
[the remark this answers is not in the document]
Comment (joe-bowman) [reply x/interchainstaking/keeper/receipt.go:77]: @coderabbitai you are correct, but until we have verified this is the case, this logic does no harm existing here. belt and braces :)

File: x/interchainstaking/keeper/receipt.go [hunk 6]
```diff
@@ -63,23 +63,32 @@
...
-		// That is okay, we'll catch it next round!)
+	valid, matchesVals := zone.ValidateCoinsForZone(assets, k.GetValidatorAddressesAsMap(ctx, zone.ChainId))
+
+	if !valid {
 		k.Logger(ctx).Error("unable to validate coins. Ignoring.", "senderAddress", senderAddress)
-		return fmt.Errorf("unable to validate coins. Ignoring. senderAddress=%q", senderAddress)
+		k.NewCompletedReceipt(ctx, &zone, senderAddress, hash, assets) // nil receipt will stop this hash being submitted again
+		// send tokens to withdrawal for disbursal.
+		return k.SendToWithdrawal(ctx, &zone, zone.DepositAddress, assets)
+	} else if !matchesVals {
```
[the remark this answers is not in the document]
Comment (joe-bowman) [reply x/interchainstaking/keeper/receipt.go:87]: @coderabbitai The invalid tokens means that this tx can never be processed via the normal flow; so NewCompletedReceipt creates a receipt marked as completed (vs NewReceipt which does not mark completed and waits for the processing to update as completed) such that we will ignore this tx in future.

File: x/interchainstaking/keeper/receipt.go [hunk 7]
```diff
@@ -63,23 +63,32 @@
...
+		k.Logger(ctx).Error("unable to validate coins for this valset.", "senderAddress", senderAddress)
+		// Do not set a nil receipt so we can revisit this tx.
+		// Don't return an error as to not clog queue.
+		return nil
```
[the remark this answers is not in the document]
Comment (joe-bowman) [reply x/interchainstaking/keeper/receipt.go:91]: @coderabbitai The absence of a nil receipt means that we will reattempt to process this tx in the future. The coins were of a valid format, but we do not yet have the data about the denom to be able to handle it. Deferring the handling of this tx until a later iteration is safe.

**Comment by faddat** [comment]
Need to check if this is still failing tests, when I first checked, it was.

**Comment by faddat** [comment]
tests fail only on amd64 darwin.  Will upgrade repo to go v1.22.0 and see if that changes this.  Seems like a compatibility issue.

**Comment by faddat** [comment]
I'm working to get the tests passing on Darwin, when they do, I'll merge this.

**Comment by faddat** [comment]
Well, that did not work.

@joe-bowman I'd say we have two choices here:

* stop testing on amd64 for darwin
* refactor TestHandleReceiptTransactionBadDenom

@coderabbitai - any reason that TestHandleReceiptTransactionBadDenom works on all platforms except for darwin/amd64?  This is a mystery to me.

File: x/interchainstaking/keeper/callbacks.go [hunk 8]
```diff
@@ -653,7 +659,9 @@
 	result := banktypes.QueryAllBalancesResponse{}
 	k.cdc.MustUnmarshal(args, &result)
 
-	zone.WithdrawalWaitgroup--
+	if err := zone.DecrementWithdrawalWaitgroup(k.Logger(ctx), 1, "delegationaccountbalances callback"); err != nil {
+		return err
+	}
```
[the remark this answers is not in the document]
Comment (joe-bowman) [reply x/interchainstaking/keeper/callbacks.go:664]: @coderabbitai - and similar to the previous comments, DecrementWithdrawalWaitgroup is a helper function that handles the potential underflow.

**Comment by faddat** [comment]
so, that was a very rigorous way to handle this... :)

**Review by faddat — APPROVED** [review]

**Pull request merged**